### ISSUE TITLE
Python error: Segmentation fault

### ISSUE THREAD

**Issue by kingosticks** [pimoroni/mopidy-pidi#3]
I've found that after running for some time, Mopidy will crash with a segmentation fault. 

> Nov 25 05:46:58 zero systemd[1]: mopidy.service: Main process exited, code=killed, status=11/SEGV
Nov 25 05:46:58 zero systemd[1]: mopidy.service: Failed with result 'signal'.

I'm not 100% sure what's causing this but I don't recall any previous reports of this from Mopidy users so I think it's possibly related to mopidy-pidi. I've not worked out a quick, reliable method to produce it yet but after leaving my Pi Zero running last night I got the following stack trace which shows mopidy-pidi as the current thread during the crash - http://dpaste.com/1XYE953.

**Comment by sandyjmacdonald** [comment]
I’ve seen the same on a Pi 3A+

**Comment by Gadgetoid** [comment]
This is going to be fun to debug. I would be interested to see if the updates to `pidi-display-st7789` and `pidi-display-pil` cause this segfault to be deferred for uh... some more time.

**Comment by kingosticks** [comment]
I'll try again with the new version and I'll also try to get a proper core dump this time. My first guess is there're a few threads going on and not much locking but I've not looked into anything in particular.

**Comment by kingosticks** [comment]
I'm not able to reproduce the crash with the latest development versions of the plugins. I think that's most likely due to how much less they now access the hardware. But I did get two other types of tracebacks  a couples of times each:

1. `IOError` - https://gist.github.com/kingosticks/37f79d70be65265deea63db43e756f12
2. `UnicodeEncodeError` - https://gist.github.com/kingosticks/199c57a8db2afaddccc4e86026cf5433

**Comment by Gadgetoid** [comment]
Thanks- you're clearly onto a couple of bugs there. I think #4 may all but fix them both, since it drastically reduces PiDi's reliance on Brainz but it's clear that handling of titles is broken and needs some TLC.

I haven't been able to reproduce the segmentation faults either, but I've had reports of similar issues with other projects drawing to similar displays. I'm relatively convinced it's a - possibly even fixed - issue in the version of PIL currently available on the Pi. I can't claim to have much experience tracking down these kinds of bugs so I'm very much hoping that it quietly goes away.

**Comment by Gadgetoid** [comment]
Have managed to reproduce a segfault in an isolated case using the following test script:

```python
from PIL import Image
import ST7735
import time


# Create LCD instance
disp = ST7735.ST7735(
    port=0,
    cs=1,
    dc=9,
    backlight=12,
    rotation=270,
    spi_speed_hz=1000000000
)

# Initialize display
disp.begin()
WIDTH = disp.width
HEIGHT = disp.height

image = Image.new('RGB', (WIDTH, HEIGHT), color=(255, 0, 255))

count = 0

while True:
    disp.display(image)
    count += 1
    if count % 1000 == 0:
        print(time.time(), count)
```

A couple of test runs on a 4GB Pi 4 show this segfaulting after count outputs 167000. Since this test script involves no PIL code whatsoever, it points the finger instead at the ST7735 library.

**Comment by Gadgetoid** [comment]
Right now I've run a test against the latest version of spidev from pypi (3.4) and have simulated about 16 days of 1-second interval updates with no segfault. This is additionally using `xfer3` instead of `xfer` which keeps the CS pin asserted between data bytes (allegedly, albeit in practise spidev is hopeless with CS) but, more importantly, takes an arbitrary length buffer and splits it into 4096 byte chunks for transfer so it's a teeny tiny bit faster.

I think it's clear this segmentation fault is originating from the st7735 library so I have raised pimoroni/st7735-python#6 to track this.

**Comment by Gadgetoid** [comment]
The segfault was rooted in spidev, and a new package has been released to Raspbian (apt) - `20190221~182651-1` - which should fix it.

**Comment by kingosticks** [comment]
Awesome! thanks.
